Hashless assets

ealmloff opened the pull request:
Closes #4115

ealmloff commented on packages/manganis/manganis-core/src/options.rs:
```diff
@@ -17,9 +17,73 @@ use crate::{
     serde::Deserialize,
 )]
+#[non_exhaustive]
+pub struct AssetOptions {
```
Changing this from an enum to a struct is somewhat breaking, but most users are probably using the variants directly like this instead of manually using the enum, so I don't expect this to break much user code:
```rust
const _: Asset = asset!("/assets/image.png", ImageAssetOptions::new().with_size(ImageSize::Manual { width: 52, height: 52 }));
```

jkelleyrtp commented:
Haven't looked at the code, but do you think it's possible to have `external_asset` be just a flag on the regular asset builder options?

ealmloff commented:
> Haven't looked at the code, but do you think it's possible to have `external_asset` be just a flag on the regular asset builder options?

No, there is a flag that controls if the hash is added, but it can't modify if the symbol is kept. `external_asset` just enables that flag and adds a `#[used]` attribute to the static

ealmloff commented:
I guess another API we could support is this with the `#[used]` in user code:
```rust
#[used]
static _EXTERNAL_ASSET: Asset = asset!(
    "/66571940.jpeg",
    ImageAssetOptions::new()
        .into_asset_options()
        .with_hash_suffix(false)
);
```

ealmloff commented:
> I beleive this happens with `dx bundle --release` on 0.7.0-alpha.1
> 
> ![image](https://private-user-images.githubusercontent.com/54854447/458912819-41f64d17-b25d-4cc3-8300-5cc43fae148d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA4NjEzMTMsIm5iZiI6MTc1MDg2MTAxMywicGF0aCI6Ii81NDg1NDQ0Ny80NTg5MTI4MTktNDFmNjRkMTctYjI1ZC00Y2MzLTgzMDAtNWNjNDNmYWUxNDhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI1VDE0MTY1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlODJkM2U5ZWQ2Nzc3ODBkMjM2NGUwNDA4ZGViZWYwOWNjYzE5ZGVjNGM5NWJmYWI2OTVjMTkxZWE3Y2Q0MzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jBlFM0hOrCouoJ9RbGbVRa4hm6Cm6ypp18h_EOD4PuE)

That placeholder is only generated by the git version of dioxus web. If you are using the `0.7.0-alpha.1` version of the CLI with the git version of dioxus web that could be the issue

ealmloff commented:
> From what I remember this might have issues with LTO since the static won't escape the module. Only extern "C" symbols make it out if they're private, but that also has some complications with duplicate symbols being defined during some edge cases in version unification.

It looks like that is a general caveat to the used attribute [in the reference](https://doc.rust-lang.org/nightly/reference/abi.html#the-used-attribute). Since we now do the asset intercept after the linker, the linker can modify/remove symbols.

In practice, this doesn't seem to be the case on macos with the default linker. If you mark the asset symbol as used, it is kept even if it is private and unused in a library as long as that library is built. `extern "C"` functions can't use the used attribute and don't show up in the symbol table after lto if you objdump the final binary. This is the setup I used to test: https://github.com/ealmloff/linker-test-kept

> Note: this isnt an "unused" image but a used one
> 
> I beleive this happens with `dx bundle --release` on 0.7.0-alpha.1
> 
> ![image](https://private-user-images.githubusercontent.com/54854447/458912819-41f64d17-b25d-4cc3-8300-5cc43fae148d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA4NjIwMTgsIm5iZiI6MTc1MDg2MTcxOCwicGF0aCI6Ii81NDg1NDQ0Ny80NTg5MTI4MTktNDFmNjRkMTctYjI1ZC00Y2MzLTgzMDAtNWNjNDNmYWUxNDhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI1VDE0MjgzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyNzI4YWJmM2NmNDZlNDNhNmZiNDhmMGJhNGRlZDM1ZDhlYTk3ZjRkM2I5OGFlYWQ5NzEzZDMxNTljMTQ5ZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.awxLvPw8t1za040eA5E__LtuLLsY11QDvjup3Y3PcFo)

I can reproduce this issue in the alpha, but not with the git version of the CLI + dioxus. It might have already been fixed?

Building on web with this profile: 

```toml
[profile.release]
opt-level = "z"
debug = false
lto = true
codegen-units = 1
panic = "abort"
strip = true
incremental = false
```

ealmloff commented:
> @ealmloff got some very bad news in this situation. Idk that it possible to force __LINK_SECTION static to persist in `dx bundle --release` in a desirable way
> 
> I cloned dioxus repo, in asset macro, I added `#[used]` .. I added `pub` .. seems the only way to get it to persist is to have __LINK_SECTION actually used in the code somewhere outside the scoping of the macro expansion
> 
> I beat the various LLMs but they all telling me there isnt another way
> 
> Here is my dioxus repo with the used n pub https://github.com/Distortedlogic/dioxus/tree/asset-link-section-preservation
> 
> Here is the repo I am bundling https://github.com/Summit-Sailors/developer-portfolio
> 
> I did fork dioxus-free-icons and bumped the dioxus version in it as I was using it in the site but thats the only dep that lso deps on dioxus https://github.com/Distortedlogic/dioxus-free-icons
> 
> > I can reproduce this issue in the alpha, but not with the git version of the CLI + dioxus. It might have already been fixed?
> 
> Did u use the --release flag with dx bundle? --release isnt default now for dx bundle .. without release works for me but also has the dev server trying to run .. with --release does not work for me even with latest git commit of dioxus
> 
> here is it working without --release ![image](https://private-user-images.githubusercontent.com/54854447/459482791-2d88f807-9280-48b9-b764-6a8340fdb96d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA5NTMwNTgsIm5iZiI6MTc1MDk1Mjc1OCwicGF0aCI6Ii81NDg1NDQ0Ny80NTk0ODI3OTEtMmQ4OGY4MDctOTI4MC00OGI5LWI3NjQtNmE4MzQwZmRiOTZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI2VDE1NDU1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyYzVhN2I2Y2FiNDdhM2E0MjI2ZmYwMDhiYzIzMWEwNGNmZTRhMzIyMWFmNGI3NGUyYTBjNzNhYjI1NDk2NWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0E9qjRuEZQ50XHbYSe32H5vB6BVERcbYtJD_QTQTTNQ)



> @ealmloff got some very bad news in this situation. Idk that it possible to force __LINK_SECTION static to persist in `dx bundle --release` in a desirable way
> 
> I cloned dioxus repo, in asset macro, I added `#[used]` .. I added `pub` .. seems the only way to get it to persist is to have __LINK_SECTION actually used in the code somewhere outside the scoping of the macro expansion
> 
> I beat the various LLMs but they all telling me there isnt another way
> 
> Here is my dioxus repo with the used n pub https://github.com/Distortedlogic/dioxus/tree/asset-link-section-preservation
> 
> Here is the repo I am bundling https://github.com/Summit-Sailors/developer-portfolio
> 
> I did fork dioxus-free-icons and bumped the dioxus version in it as I was using it in the site but thats the only dep that lso deps on dioxus https://github.com/Distortedlogic/dioxus-free-icons
> 
> > I can reproduce this issue in the alpha, but not with the git version of the CLI + dioxus. It might have already been fixed?
> 
> Did u use the --release flag with dx bundle? --release isnt default now for dx bundle .. without release works for me but also has the dev server trying to run .. with --release does not work for me even with latest git commit of dioxus
> 
> here is it working without --release ![image](https://private-user-images.githubusercontent.com/54854447/459482791-2d88f807-9280-48b9-b764-6a8340fdb96d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA5NTMwNTgsIm5iZiI6MTc1MDk1Mjc1OCwicGF0aCI6Ii81NDg1NDQ0Ny80NTk0ODI3OTEtMmQ4OGY4MDctOTI4MC00OGI5LWI3NjQtNmE4MzQwZmRiOTZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI2VDE1NDU1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyYzVhN2I2Y2FiNDdhM2E0MjI2ZmYwMDhiYzIzMWEwNGNmZTRhMzIyMWFmNGI3NGUyYTBjNzNhYjI1NDk2NWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0E9qjRuEZQ50XHbYSe32H5vB6BVERcbYtJD_QTQTTNQ)

You have `strip = true` defined in your cargo.toml which is not compatible with the symbol based asset system when building native binaries like the server. Try removing it or only enabling it in web builds:

https://github.com/Summit-Sailors/developer-portfolio/blob/main/Cargo.toml#L26

Distortedlogic commented:
yes, I tried without strip and i tried without lto

with strip = false
```
44.629s DEBUG dx::build::request: Build completed successfully in 44349857us: "/home/entropybender/repos/developer-portfolio/target/wasm32-unknown-unknown/release/app.wasm"
  44.629s  INFO dx::build::builder: Bundling app...
  44.629s DEBUG dx::build::request: Running wasm-bindgen dx_src=bundle
  44.629s  INFO dx::build::builder: Running wasm-bindgen...
  44.629s DEBUG dx::wasm_bindgen: wasm-bindgen: [
    "--target",
    "web",
    "--debug",
    "--no-demangle",
    "--keep-debug",
    "--keep-lld-exports",
    "--out-name",
    "app",
    "--no-typescript",
    "--out-dir",
    "/home/entropybender/repos/developer-portfolio/target/dx/app/release/web/public/wasm",
    "/home/entropybender/repos/developer-portfolio/target/wasm32-unknown-unknown/release/app.wasm",
]
  44.931s DEBUG dx::build::request: wasm-bindgen complete in 301.972793ms dx_src=bundle
  44.982s DEBUG dx::build::builder: Setting builder to failed state
  44.982s ERROR dx::build::builder: err=Other(Failed to write main executable

Caused by:
    0: Failed to read module
    1: [parse exception: invalid code after misc prefix: 17 (at 0:559040)]) 
Failed to write main executable
```

with lto=false it builds but issue persists

with strip=false and lto=false

```
39.731s  INFO dx::build::builder: Bundling app...
  39.731s DEBUG dx::build::request: Running wasm-bindgen dx_src=bundle
  39.731s  INFO dx::build::builder: Running wasm-bindgen...
  39.731s DEBUG dx::wasm_bindgen: wasm-bindgen: [
    "--target",
    "web",
    "--debug",
    "--no-demangle",
    "--keep-debug",
    "--keep-lld-exports",
    "--out-name",
    "app",
    "--no-typescript",
    "--out-dir",
    "/home/entropybender/repos/developer-portfolio/target/dx/app/release/web/public/wasm",
    "/home/entropybender/repos/developer-portfolio/target/wasm32-unknown-unknown/release/app.wasm",
]
  40.226s DEBUG dx::build::request: wasm-bindgen complete in 495.435904ms dx_src=bundle
  40.283s DEBUG dx::build::builder: Setting builder to failed state
  40.283s ERROR dx::build::builder: err=Other(Failed to write main executable

Caused by:
    0: Failed to read module
    1: [parse exception: invalid code after misc prefix: 17 (at 0:662914)]) 
Failed to write main executable
```

I tried a few other various combos and the doixosu toml wasm opt settings :/

ealmloff commented on packages/manganis/manganis-core/src/css.rs:
```diff
@@ -21,35 +21,60 @@ pub struct CssAssetOptions {
     /// Create a new css asset using the builder
-    pub const fn new() -> Self {
+    #[deprecated(note = "Use `AssetOptions::css()` instead", since = "0.7.0")]
+    pub const fn new() -> AssetOptionsBuilder<CssAssetOptions> {
```
On the fence about deprecating this. It no longer returns Self which is odd, but this is probably what you want if you try to build `CssAssetOptions`. Keeping the new method makes this change much less breaking
